src/backend/distributed/planner/multi_planner.c

```diff
@@ -59,6 +59,26 @@ multi_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
 	if (needsDistributedPlanning)
 	{
 		originalQuery = copyObject(parse);
+
+		/*
+		 * We implement INSERT INTO .. SELECT by pushing down the SELECT to
+		 * each shard. To compute that we use the router planner, by adding
+		 * an "uninstantiated" constraint that the partition column be equal to a
+		 * certain value. standard_planner() distributes that constraint to
+		 * the baserestrictinfos of the tables that they are connected
+		 * via equi joins.
```

Comment:
s/connected via equi joins/to all tables where it knows how to push the restriction safely/? Because this will work for some other things than equijoins...